x-pack/plugins/reporting/server/browsers/phantom/driver/index.js

```diff
@@ -18,38 +18,87 @@ export function PhantomDriver({ page, browser, zoom, logger }) {
     if (page === false || browser === false) throw new Error('Phantom instance is closed');
   };
 
-  const configurePage = () => {
+  const configurePage = (pageOptions) => {
     const RESOURCE_TIMEOUT = 5000;
     return fromCallback(cb => page.set('resourceTimeout', RESOURCE_TIMEOUT, cb))
       .then(() => {
         if (zoom) return fromCallback(cb => page.set('zoomFactor', zoom, cb));
+      })
+      .then(() => {
+        if (pageOptions.conditionalHeaders) {
+          const headers = pageOptions.conditionalHeaders.headers;
+          const conditions = pageOptions.conditionalHeaders.conditions;
+
+          const escape = (str) => {
+            return str
+              .replace(/'/g, `\\'`)
+              .replace(/\\/g, `\\\\`)
+              .replace(/\r?\n/g, '\\n');
+          };
+
+          // we're using base64 encoding for any user generated values that we need to eval
+          // to be sure that we're handling these properly
+          const btoa = (str) => {
+            return Buffer.from(str).toString('base64');
+          };
+
+          const fn = `function (requestData, networkRequest) {
+            var log = function (msg) {
+              if (!page.onConsoleMessage) {
+                return;
+              }
+              page.onConsoleMessage(msg);
+            };
+
+            var parseUrl = function (url) {
+              var link = document.createElement('a');
+              link.href = url;
+              return {
+                protocol: link.protocol,
+                port: link.port,
+                hostname: link.hostname,
+                pathname: link.pathname,
+              };
+            };
+
+            var shouldUseCustomHeadersForPort = function (port) {
+              if ('${escape(conditions.protocol)}' === 'http' && ${conditions.port} === 80) {
+                return port === undefined || port === null || port === '' || port === '${conditions.port}';
```

Comment:
Why does this port check (and the `443` version below) allow for empty string, but the chromium version does not? Is this a side effect of the way we parse for the port in each scenario?

Reply:
Yup, when using the `URL` class directly, we get a `null` for the port, where-as our parsing logic in PhantomJS gives us an empty string in these scenarios.

```
var url = require('url');
url.parse('https://google.com');
Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'google.com',
  port: null,
  hostname: 'google.com',
  hash: null,
  search: null,
  query: null,
  pathname: '/',
  path: '/',
  href: 'https://google.com/' }
```

I'm tempted to include the empty string check here as well, since doing the following in a recent version of Chrome gives us an empty string:
```
new URL('https://google.com');
URL {href: "http://google.com/", origin: "http://google.com", protocol: "http:", username: "", password: "", …}
hash: ""
host: "google.com"
hostname: "google.com"
href: "http://google.com/"
origin: "http://google.com"
password: ""
pathname: "/"
port: ""
protocol: "http:"
search: ""
searchParams: URLSearchParams {}
username: ""
```

Reply:
Thanks for the explanation! I agree an empty string check would be worthwhile for chromium too.